### PR TITLE
Create delete resource and remove finalizer functions.

### PR DESCRIPTION
**Pull request by tbeerbower**
There is a lot of repeated code to maintain throughout the uninstaller. This code can be refactored to make maintaining and updating easier. 
The code should transform:
```
kubectl get csr --no-headers -o custom-columns=":metadata.name" \
    | awk '/csr-/' \
    | xargsr kubectl delete csr \
    || return $?
```
to this:
```
delete_k8s_resources csr '/csr-/' ":metadata.name" || return $?
```

### CODE REVIEW

File: uninstall/uninstall-steps/1-uninstall-istio.sh [hunk 1]
```diff
@@ -60,17 +52,12 @@ function delete_secrets() {
...
-  kubectl get secrets -n kube-system --no-headers -o custom-columns=":metadata.name,:metadata.annotations" \
-  | awk '/istio./ {print $1}' \
-  | xargsr kubectl delete secret -n kube-system \
-  || err_return $? "Could not delete secrets from Istio in namespace kube-system" || return $? # return on pipefail
+  delete_k8s_resources secrets ":metadata.name,:metadata.annotations" '/istio./ {print $1}' "kube-system" \
+    || err_return $? "Could not delete secrets from Istio in namespace kube-system" || return $? # return on pipefail
 }
 
 function delete_istio_namepsace() {
-  log "Deleting istio-system finalizers"
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/1-uninstall-istio.sh:70]: This log message should not have been deleted.

File: uninstall/uninstall-steps/1-uninstall-istio.sh [hunk 2]
```diff
@@ -79,15 +66,10 @@ function delete_istio_namepsace() {
 
 function finalize() {
   # Removing possible reference to verrazzano in clusterroles and clusterrolebindings
-  log "Removing Verrazzano ClusterRoles and ClusterRoleBindings"
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/1-uninstall-istio.sh:82]: Same as previous comment

File: uninstall/uninstall-utils.sh [hunk 3]
```diff
@@ -42,4 +42,32 @@ function err_return() {
...
+# $2 custom-cols   - custom columns used when getting the resources
+# $3 filter-args   - arguments given to awk to filter the resource list for deletion (optional)
+# $4 namespace     - namespace for resources to be deleted (optional)
+# Usage:
+# delete_k8s_resources resource-type custom-cols [filter-args] [namespace]
+function delete_k8s_resources() {
+  ( [ -z "$4" ] && kubectl get $1 --no-headers -o custom-columns="$2" || kubectl get $1 --no-headers -o custom-columns="$2" -n "$4" ) \
+    | ( [ -z "$3" ] && cat || awk "$3" ) \
+    | ( [ -z "$4" ] && xargsr kubectl delete $1 || xargsr kubectl delete $1 -n "$4" ) \
+    || return $? # return on pipefail
```
Comment (davis-gomes) [inline uninstall/uninstall-utils.sh:58]: This could utilize the err_return function.

File: uninstall/uninstall-utils.sh [hunk 4]
```diff
@@ -42,4 +42,32 @@ function err_return() {
...
+# $1 resource-type - type of the resources being patched
+# $2 custom-cols   - custom columns used when getting the resources
+# $3 filter-args   - arguments given to awk to filter the resource list for patching
+# $4 patch         - the patch to be applied to the resources
+# Usage:
+# patch_k8s_resources resource-type custom-cols filter-args patch
+function patch_k8s_resources() {
+  kubectl get $1 --no-headers -o custom-columns="$2" \
+    | awk "$3" \
+    | xargsr kubectl patch $1 -p "$4" --type=merge \
```
Comment (davis-gomes) [inline uninstall/uninstall-utils.sh:71]: Will this return an error if $4 is empty?

File: uninstall/uninstall-steps/2-uninstall-system-components.sh [hunk 5]
```diff
@@ -93,16 +89,11 @@ function delete_rancher() {
...
   do
     # remove finalizers from crds
-    kubectl get crds --no-headers -o custom-columns=":metadata.name,:spec.group" \
-      | awk '/coreos.com|cattle.io/ {print $1}' \
-      | xargsr kubectl patch crd -p '{"metadata":{"finalizers":null}}' --type=merge \
+    patch_k8s_resources crds ":metadata.name,:spec.group" '/coreos.com|cattle.io/ {print $1}' '{"metadata":{"finalizers":null}}' \
       || err_return $? "Could not remove finalizers from CustomResourceDefinitions in Rancher" || return $? # return on pipefail
 
     # delete crds
-    # This process is backgrounded in order to timeout due to finalizers hanging
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/2-uninstall-system-components.sh:102]: Please leave this comment in. It is a reminder for future edits to the uninstall.

File: uninstall/uninstall-steps/4-uninstall-keycloak.sh [hunk 6]
```diff
@@ -30,10 +30,7 @@ function delete_keycloak() {
     || err_return $? "Could not delete keycloak from helm" || return $? # return on pipefail
 
   # delete keycloak namespace
-  log "Deleting keycloak namespace finalizers"
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/4-uninstall-keycloak.sh:33]: Include this log message.

File: uninstall/uninstall-steps/3-uninstall-verrazzano.sh [hunk 7]
```diff
@@ -26,50 +26,34 @@ function delete_verrazzano() {
...
   # deleting namespaces
   log "Deleting Verrazzano namespace finalizers"
   # delete namespace finalizers
-  kubectl get namespace --no-headers -o custom-columns=":metadata.name,:metadata.labels" \
-    | awk '/k8s-app:verrazzano.io|verrazzano-system/ {print $1}' \
-    | xargsr kubectl patch namespace -p '{"metadata":{"finalizers":null}}' --type=merge \
+  patch_k8s_resources namespace ":metadata.name,:metadata.labels" '/k8s-app:verrazzano.io|verrazzano-system/ {print $1}' '{"metadata":{"finalizers":null}}' \
     || err_return $? "Could not remove finalizers from Verrazzano namespaces" || return $? # return on pipefail
 
-  log "Deleting Verrazzano namespaces"
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/3-uninstall-verrazzano.sh:69]: This one as well.

File: uninstall/uninstall-steps/3-uninstall-verrazzano.sh [hunk 8]
```diff
@@ -26,50 +26,34 @@ function delete_verrazzano() {
...
   log "Deleting ClusterRoles and ClusterRoleBindings"
   # deleting clusterrolebindings
-  kubectl get clusterrolebinding --no-headers -o custom-columns=":metadata.name,:metadata.labels" \
-    | awk '/verrazzano/ {print $1}' \
-    | xargsr kubectl delete clusterrolebinding \
+  delete_k8s_resources clusterrolebinding ":metadata.name,:metadata.labels" '/verrazzano/ {print $1}' \
     || err_return $? "Could not delete ClusterRoleBindings from Verrazzano" || return $? # return on pipefail
 
   # deleting clusterroles
-  log "Deleting ClusterRoles"
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/3-uninstall-verrazzano.sh:55]: Include this as well.

File: uninstall/uninstall-steps/1-uninstall-istio.sh (hunk 1 above)
Comment (tbeerbower) [reply uninstall/uninstall-steps/1-uninstall-istio.sh:70]: Sorry about the dropped log messages.  It happened during the rebase.  Thanks for catching it.

File: uninstall/uninstall-utils.sh (hunk 3 above)
Comment (tbeerbower) [reply uninstall/uninstall-utils.sh:58]: In all cases the err_return is called by the caller of this function.  Are you suggesting to add another call to err_return here or add the error message as an argument to this function and only call it here?  I guess it makes sense to pass the error message in and call it here.  I just picked up the error messages on the rebase so I hadn't considered it earlier.

File: uninstall/uninstall-utils.sh (hunk 4 above)
Comment (tbeerbower) [reply uninstall/uninstall-utils.sh:71]: $4 is not optional.  It shouldn't be empty.  Non of the current usages pass an empty patch.

File: uninstall/uninstall-utils.sh (hunk 3 above)
Comment (davis-gomes) [reply uninstall/uninstall-utils.sh:58]: I guess you're right in that every error here would result in an error message output. You could consider including it inside the function. I am fine with whichever makes more intuitive sense to you.

File: uninstall/uninstall-utils.sh (hunk 4 above)
Comment (davis-gomes) [reply uninstall/uninstall-utils.sh:71]: I am worried that if, in the future, this function is used and the argument is left out, that patching will "fail" silently and not patch any finalizers.

Comment (tbeerbower) [reply uninstall/uninstall-utils.sh:71]: I'll add a verification.

File: uninstall/uninstall-utils.sh (hunk 3 above)
Comment (tbeerbower) [reply uninstall/uninstall-utils.sh:58]: I'm adding it to the function.  I think it makes sense but the arg list is getting long which makes it a bit harder to read.  But the goal was to reduce duplicated code so I'll move the err_return into the functions.

File: uninstall/uninstall-steps/1-uninstall-istio.sh (hunk 1 above)
Comment (tbeerbower) [reply uninstall/uninstall-steps/1-uninstall-istio.sh:70]: @davis-gomes I think that I have all the log messages back in place.

File: uninstall/uninstall-utils.sh (hunk 3 above)
Comment (tbeerbower) [reply uninstall/uninstall-utils.sh]: I added the err_return call the the functions.  The message is passed in by the caller.

File: uninstall/uninstall-utils.sh (hunk 4 above)
Comment (tbeerbower) [reply uninstall/uninstall-utils.sh]: I added a check for the patch arg.

File: uninstall/uninstall-steps/2-uninstall-system-components.sh [hunk 9]
```diff
@@ -140,17 +126,12 @@ function delete_rancher() {
...
   # delete namespace finalizers
-  kubectl get namespaces --no-headers -o custom-columns=":metadata.name" \
-    | awk '/cattle-|local|p-|user-/ {print $1}' \
-    | xargsr kubectl patch namespace -p '{"metadata":{"finalizers":null}}' --type=merge \
-    || err_return $? "Could not remove finalizers from namespaces in Rancher" || return $? # return on pipefail
+  patch_k8s_resources namespaces ":metadata.name" "Could not remove finalizers from namespaces in Rancher" '/cattle-|local|p-|user-/ {print $1}' '{"metadata":{"finalizers":null}}' \
+    || return $? # return on pipefail
 
   # delete cattle namespaces
-  log "Delete rancher namespace"
```
Comment (davis-gomes) [inline uninstall/uninstall-steps/2-uninstall-system-components.sh:149]: I missed this one.

Comment (tbeerbower) [reply uninstall/uninstall-steps/2-uninstall-system-components.sh:149]: @davis-gomes Thanks!  Added back.

**Review by davis-gomes — APPROVED** [review]
I just tested it, and it looks good!

**Pull request merged**